ompi/mca/coll/acoll/coll_acoll_utils.h

```diff
@@ -0,0 +1,769 @@
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    data->xpmem_raddr = (void **) malloc(sizeof(void *) * size);
+    if (NULL == data->xpmem_raddr) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    data->rcache = (mca_rcache_base_module_t **) malloc(sizeof(mca_rcache_base_module_t *) * size);
+    if (NULL == data->rcache) {
+        line = __LINE__;
+        ret = OMPI_ERR_OUT_OF_RESOURCE;
+        goto error_hndl;
+    }
+    seg_id = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE, (void *) 0666);
+    if (seg_id == -1) {
+        line = __LINE__;
+        ret = -1;
+        goto error_hndl;
+    }
+
+    ret = comm->c_coll->coll_allgather(&seg_id, sizeof(xpmem_segid_t), MPI_BYTE, data->allseg_id,
+                                       sizeof(xpmem_segid_t), MPI_BYTE, comm,
+                                       comm->c_coll->coll_allgather_module);
+
+    /* Assuming the length of rcache name is less than 50 characters */
+    char rc_name[50];
+    for (int i = 0; i < size; i++) {
+        if (rank != i) {
+            data->all_apid[i] = xpmem_get(data->allseg_id[i], XPMEM_RDWR, XPMEM_PERMIT_MODE,
+                                          (void *) 0666);
+            if (data->all_apid[i] == -1) {
+                line = __LINE__;
+                ret = -1;
+                goto error_hndl;
+            }
+            if (data->all_apid[i] == -1) {
+                line = __LINE__;
+                ret = -1;
+                goto error_hndl;
+            }
+            sprintf(rc_name, "acoll_%d_%d_%d", cid, rank, i);
+            mca_rcache_base_resources_t rcache_element
+                = {.cache_name = rc_name,
+                   .reg_data = &data->all_apid[i],
+                   .sizeof_reg = sizeof(struct acoll_xpmem_rcache_reg_t),
+                   .register_mem = mca_coll_acoll_xpmem_register,
+                   .deregister_mem = mca_coll_acoll_xpmem_deregister};
+
+            data->rcache[i] = mca_rcache_base_module_create("grdma", NULL, &rcache_element);
+            if (data->rcache[i] == NULL) {
+                ret = -1;
+                line = __LINE__;
+                goto error_hndl;
+            }
+        }
+    }
+#endif
+
+    /* temporary variables */
+    int tmp1, tmp2, tmp3 = 0;
+    comm_grp_ranks_local(comm, subc->numa_comm, &tmp1, &tmp2, &data->l1_gp, tmp3);
+    data->l1_gp_size = ompi_comm_size(subc->numa_comm);
+    data->l1_local_rank = ompi_comm_rank(subc->numa_comm);
+
+    comm_grp_ranks_local(comm, subc->numa_comm_ldrs, &tmp1, &tmp2, &data->l2_gp, tmp3);
+    data->l2_gp_size = ompi_comm_size(subc->numa_comm_ldrs);
+    data->l2_local_rank = ompi_comm_rank(subc->numa_comm_ldrs);
+    data->offset[0] = 16 * 1024;
+    data->offset[1] = data->offset[0] + size * 64;
+    data->offset[2] = data->offset[1] + size * 64;
+    data->offset[3] = data->offset[2] + rank * 8 * 1024;
```

Comment:
Where do these magic numbers come from? What offsets do they encode?

Reply:
Updated in the lastest push.